File: entities/explosion.py
"""
DY01 - 爆炸
"""
import pygame


class Explosion:
    def __init__(self, x, y, radius, damage):
        self.x = x
        self.y = y
        self.radius = radius
        self.damage = damage
        self.life = 22
        self.max_life = 22

    def update(self):
        self.life -= 1
        return self.life > 0

    def draw(self, surf, cam_x):
        a = self.life / self.max_life
        r = self.radius * (1 - a * 0.5) * (a / 0.3 if a < 0.3 else 1)
        cx, cy = int(self.x - cam_x), int(self.y)

        s = pygame.Surface((int(r * 2 + 4), int(r * 2 + 4)), pygame.SRCALPHA)
        pygame.draw.circle(s, (255, 102, 0, int(a * 0.6 * 255)), (int(r + 2), int(r + 2)), int(r))
        pygame.draw.circle(s, (255, 204, 0, int(a * 0.8 * 255)), (int(r + 2), int(r + 2)), int(r * 0.55))
        pygame.draw.circle(s, (255, 255, 255, int(a * 255)), (int(r + 2), int(r + 2)), int(r * 0.25))
        surf.blit(s, (cx - int(r + 2), cy - int(r + 2)))

File: monogame/Entities/Explosion.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;

namespace DY01.Entities;

public class Explosion
{
    public float X, Y, Radius, Damage;
    public int Life = 22;
    public int MaxLife = 22;

    public Explosion(float x, float y, float radius, float damage)
    {
        X = x; Y = y; Radius = radius; Damage = damage;
    }

    public bool Update()
    {
        Life--;
        return Life > 0;
    }

    public void Draw(SpriteBatch sb, Texture2D pixel, float camX)
    {
        float a = (float)Life / MaxLife;
        float r = Radius * (1 - a * 0.5f) * (a < 0.3f ? a / 0.3f : 1);
        int cx = (int)(X - camX), cy = (int)Y;

        // Outer ring
        var c1 = new Color(255, 102, 0) * (a * 0.6f);
        DrawFilledCircle(sb, pixel, new Vector2(cx, cy), r, c1);

        // Mid ring
        var c2 = new Color(255, 204, 0) * (a * 0.8f);
        DrawFilledCircle(sb, pixel, new Vector2(cx, cy), r * 0.55f, c2);

        // Inner
        var c3 = Color.White * a;
        DrawFilledCircle(sb, pixel, new Vector2(cx, cy), r * 0.25f, c3);
    }

    private void DrawFilledCircle(SpriteBatch sb, Texture2D pixel, Vector2 center, float radius, Color color)
    {
        int r = (int)radius;
        for (int dy = -r; dy <= r; dy++)
        {
            for (int dx = -r; dx <= r; dx++)
            {
                if (dx * dx + dy * dy <= r * r)
                {
                    sb.Draw(pixel, new Rectangle((int)center.X + dx, (int)center.Y + dy, 1, 1), color);
                }
            }
        }
    }
}